cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/main/MessageUtil.java

```diff
@@ -178,35 +176,41 @@ private static long getTime(String runId, TimeUnit timeUnit) {
     return Collections.emptyList();
   }
 
-  private static String getLatestMessageIdFromFile(Location latest) throws InterruptedException {
-    boolean success = false;
+  private static String getLatestMessageId(List<Location> locationsTimeSorted) throws InterruptedException {
     String messageId = null;
-    while (!success) {
+    int index = locationsTimeSorted.size() - 1;
+    // there's a possibility that the latest file could be empty, so we go through the list of locations in
+    // descending time order and return the messageId from the latest file which has content.
+    while (index >= 0) {
+      Location latestLocation = locationsTimeSorted.get(index);
       try {
         DataFileReader<GenericRecord> dataFileReader =
-          new DataFileReader<>(new File(latest.toURI()),
-                                            new GenericDatumReader<>(ProgramRunIdFieldsSerializer.SCHEMA));
-        // able to get data file reader without any IOException
-        success = true;
+          new DataFileReader<>(new File(latestLocation.toURI()),
+                               new GenericDatumReader<>(ProgramRunInfoSerializer.SCHEMA));
+        //TODO - efficiently skip using sync points CDAP-13400
         while (dataFileReader.hasNext()) {
           GenericRecord record = dataFileReader.next();
           messageId = record.get(Constants.MESSAGE_ID).toString();
         }
+        if (messageId != null) {
+          return messageId;
+        }
+        index--;
```

Comment:
if there are more than one empty files, there might be some problems, should log?

Reply:
why would there be a problem ? if there are more than one empty file, we will keep looking at the next latest file which has content, if all of the files in the namespace are empty, we will return null. which will be handled.

Reply:
nvm, I thought no empty file will be created if the current file is empty. was not aware that the files can be rotated if it's been open too long

Reply:
mostly the failure case where the file is created but before any record is written to it, the program is stopped either manually or due to exception would create empty files.